bpo-43352: Add a Barrier object in asyncio lib

YvesDup opened the pull request:
[bpo-43352](https://bugs.python.org/issue43352): Add a Barrier object in asyncio lib

Copy design of object Barrier from threading, same features

Modified files are:

file lib/asyncio/locks.py
+ Add Barrier and BrokenBarrierError objects
+ Added these 2 previous objects to '__all__' variable

file lib/test/test_asyncio/test_locks.py
+ Add a BarrierTests object
+ Modify STR_RGX_REPR to integrate '__repr__' of Barrier
+ Add a new test with Barrier in 'LockTest.test_lock_doesnt_accept_loop_parameter'

I am not familiar with git, so all files with .rst.bak extension are not in the scope of this PR

Just see https://bugs.python.org/issue37945, so the following paragraph can be deleted
"""
All tests (Win10 Python x64) are done and succesfull except one: python -m test -v test_locale 
I am surprised about this error because, asyncio and locale are not linked, or are they ?
Please, how to proceed about this failure ? Open a new issue on bpo ?
a log file is available.
Thank for help
"""


<!-- issue-number: [bpo-43352](https://bugs.python.org/issue43352) -->
https://bugs.python.org/issue43352
<!-- /issue-number -->

YvesDup committed:
Initial commit

Add a Barrier object to asyncio synchronized primitives

YvesDup committed:
Initial commit

Add a Barrier object to asyncio synchronized primitives

YvesDup committed:
make patchcheck

Change after run make patchcheck

YvesDup committed:
make patchcheck

Change after run make patchcheck

YvesDup committed:
make patchcheck

run make patchcheck

YvesDup committed:
make patchcheck

run make patchcheck

YvesDup committed:
Update test_locks.py

YvesDup committed:
Update test_locks.py

YvesDup committed:
Update locks.py

Add comment when testing self._state
Change if to while instruction  in _block method

YvesDup committed:
Update locks.py

Add comment when testing self._state
Change if to while instruction  in _block method

YvesDup committed:
Update test_locks.py

Refactoring all test names of BarrierTests to be more readable
Add a method cancel_coros
Refactoring and simplifyiing some tests
Correction of RGX_REPR

YvesDup committed:
Update test_locks.py

Refactoring all test names of BarrierTests to be more readable
Add a method cancel_coros
Refactoring and simplifyiing some tests
Correction of RGX_REPR

YvesDup committed:
Update from patch

after run python.bat Tools\scripts\patchcheck.py

YvesDup committed:
Update from patch

after run python.bat Tools\scripts\patchcheck.py

YvesDup committed:
remove cpython/doc/*.rst.bak

YvesDup committed:
remove cpython/doc/*.rst.bak

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

YvesDup committed:
Update 2021-03-31-15-22-45.bpo-43352.nSjMuE.rst

YvesDup committed:
Update 2021-03-31-15-22-45.bpo-43352.nSjMuE.rst

YvesDup committed:
Update documentation about add of Barrier object

Add  Barrier object as a clone of threading.Barrier

YvesDup committed:
Update documentation about add of Barrier object

Add  Barrier object as a clone of threading.Barrier

YvesDup committed:
Merge branch 'fix-issue-43352' of https://github.com/YvesDup/cpython into fix-issue-43352

asvetlov reviewed (commented):
Is the execution of `action` performed in a *random* task?

We can be more precise for this I believe.
For example, declare that the first task that calls `wait()` executes the action callback.

YvesDup committed:
Update asyncio-sync.rst

Describe when  action is called

YvesDup committed:
Update asyncio-sync.rst

Describe when  action is called

eamanu commented on Lib/asyncio/locks.py:
```diff
@@ -418,3 +418,174 @@ def release(self):
+# similar to 'draining' except that tasks leave with a BrokenBarrierError,
+# and a 'broken' state in which all tasks get the exception.
+
+class Barrier(mixins._LoopBoundMixin):
```
IMO you should also mention that is based on the Threading work.

YvesDup replied:
This is already mentioned in docstring class.

eamanu commented on Doc/library/asyncio-sync.rst:
```diff
@@ -325,6 +326,22 @@ BoundedSemaphore
+Barrier
+=======
+
+.. class:: Barrier(parties, action=None)
```
Please, could you explain a little more on documentation? I mean adding some methods docs, and examples?

YvesDup replied:
Ok I am working on today

YvesDup replied:
Update documentation with a sample and a complete description of the class behavior

eamanu commented on Lib/asyncio/locks.py:
```diff
@@ -1,6 +1,6 @@
 """Synchronization primitives."""
 
-__all__ = ('Lock', 'Event', 'Condition', 'Semaphore', 'BoundedSemaphore')
+__all__ = ('Lock', 'Event', 'Condition', 'Semaphore', 'BoundedSemaphore', 'Barrier', 'BrokenBarrierError')
```
take care pep8

YvesDup replied:
Done, thank

eamanu commented on Lib/asyncio/locks.py:
```diff
@@ -418,3 +418,174 @@ def release(self):
+        #see if the barrier is in a broken state
+        if self._state < 0:
+            raise BrokenBarrierError
+        assert self._state == 0, repr(self)
```
IMO assert should be used on tests. I know that `assert` is used on threading 
Barrier object, but here can we be more specific?

YvesDup replied:
Ok, this ``assert`` and this one of ``_wait`` method will be removed. 
* In ``_block`` method, it is not necessary to add a new test because ``self._state`` is always equal to 0
* In ``_wait`` method, no more tests are needed

eamanu commented on Lib/asyncio/locks.py:
```diff
@@ -418,3 +418,174 @@ def release(self):
+            self._waiting.set()
+        except:
+            #an exception during the _action handler.  Break and reraise
+            self._state = -2
```
this is similar to abort()

YvesDup replied:
Done, thank

eamanu commented on Lib/test/test_asyncio/test_locks.py:
```diff
@@ -928,5 +940,404 @@ def test_release_no_waiters(self):
+        barrier = asyncio.Barrier(self.N)
+        self.assertEqual(barrier.n_waiting, 0)
+        r1 = self.loop.run_until_complete(barrier.wait())
+        self.assertEqual(barrier.n_waiting, 0)
```
must be `self.assertEqual(barrier.n_waiting, 1)` ?

YvesDup replied:
No, because in this test, Barrier object is created with a ``parties`` parameter to 1. 
So when call ``barrier.wait()``, you never wait, you instantly **drain** 0 task.  So in this case, ``barrier.n_waiting`` is always to 0

YvesDup committed:
Update locks.py

first feedbacks from Emmanuel Arias

YvesDup committed:
Update locks.py

first feedbacks from Emmanuel Arias

YvesDup committed:
Update test_locks.py

Remove  "import functools"

YvesDup committed:
Update test_locks.py

Remove  "import functools"

YvesDup committed:
Update asyncio-sync.rst

Suggestion of Emmanuel Arias: Add sample and description about Barrier Class

YvesDup committed:
Update asyncio-sync.rst

Suggestion of Emmanuel Arias: Add sample and description about Barrier Class

YvesDup committed:
Update asyncio-sync.rst

bpo-43352 - Add a sample and description for Barrier Object

Minor updates as remove crlf

YvesDup committed:
Update asyncio-sync.rst

bpo-43352 - Add a sample and description for Barrier Object

Minor updates as remove crlf

YvesDup committed:
Update asyncio-sync.rst

bpo-43352 Suppress ` in a comment of the example

YvesDup committed:
Update asyncio-sync.rst

bpo-43352 Suppress ` in a comment of the example

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Add a class BarrierTests to test Barrier object
Run i17 tests ncluding one about Barrier does not accept loop parameter

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Add a class BarrierTests to test Barrier object
Run i17 tests ncluding one about Barrier does not accept loop parameter

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Add the Barrier object in Synchronization section
Add a link to a Barrier sample  in Examples sub-section
Add the BrokenBarrierError Exception in Exception section

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Add the Barrier object in Synchronization section
Add a link to a Barrier sample  in Examples sub-section
Add the BrokenBarrierError Exception in Exception section

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Add a complete documentation inspired from threading.Barrier
Add a sample  to show how use a Barrier Object

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Add a complete documentation inspired from threading.Barrier
Add a sample  to show how use a Barrier Object

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Add a Barrier object as a clone from thraeding Barrier
Update of __all__ variable to include this new class and the related BrokenBarrierError exception

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Add a Barrier object as a clone from thraeding Barrier
Update of __all__ variable to include this new class and the related BrokenBarrierError exception

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Add a class BarrierTests to test Barrier object
Run 17 tests including one about Barrier does not accept loop parameter

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Add a class BarrierTests to test Barrier object
Run 17 tests including one about Barrier does not accept loop parameter

YvesDup committed:
Merge remote-tracking branch 'upstream/main' into fix-issue-43352

YvesDup committed:
Merge branch 'fix-issue-43352' of github.com:YvesDup/cpython into fix-issue-43352

YvesDup committed:
Merge remote-tracking branch 'upstream/main' into fix-issue-43352

1st1 commented on Doc/library/asyncio-sync.rst:
```diff
@@ -356,6 +357,135 @@ BoundedSemaphore
+Barrier
+=======
+
+   .. versionadded:: 3.10
```
It's too late to push this to 3.10, but 3.11 is OK.

YvesDup replied:
Ok, I will change

1st1 commented on Doc/library/asyncio-sync.rst:
```diff
@@ -356,6 +357,135 @@ BoundedSemaphore
+
+   This class, as a clone of :class:`threading.Barrier`, provides a simple synchronization
+   primitive for use by a fixed number of tasks that need to wait for each other.
+   Each of the task tries to pass the barrier by calling the :meth:`~Barrier.wait` method
```
```suggestion
   Each of the tasks tries to pass the barrier by calling the :meth:`~Barrier.wait` method
```

1st1 commented on Doc/library/asyncio-sync.rst:
```diff
@@ -356,6 +357,135 @@ BoundedSemaphore
+   This class, as a clone of :class:`threading.Barrier`, provides a simple synchronization
+   primitive for use by a fixed number of tasks that need to wait for each other.
+   Each of the task tries to pass the barrier by calling the :meth:`~Barrier.wait` method
+   and will block until all of the tasks have made their :meth:`~Barrier.wait` calls.
```
```suggestion
   and would block until all of the tasks have made their :meth:`~Barrier.wait` calls.
```

1st1 commented on Doc/library/asyncio-sync.rst:
```diff
@@ -356,6 +357,135 @@ BoundedSemaphore
+.. class:: Barrier(parties, action=None)
+
+   Create a barrier object for *parties* number of tasks.  A callable *action*, when
+   provided, is called once just before the release phasis (draining state).
```
I'm wondering if the design of this primitive should deviate from the threading version. In this case, I think we should rely on `async with` instead of passing a callback (which are always a PITA to deal with in async/await context) :

```python
async with asyncio.Barrier(10) as b:
    tasks = [asyncio.create_task(remove_file(barrier, f)) for f in files]
    tasks.append(asyncio.create_task(remove_folder(barrier, folder)))
    await asyncio.gather(*tasks)

msg = 'ACTION: All files removed'
print(f'{os.linesep}{msg:-^80s}{os.linesep}')
```

YvesDup replied:
Thank you @1st1 for your response. 

Notice: when specified, the callback runs **once** from the last task calling the  `wait` method and just before draining the barrier. If the callback crashes, the barrier state goes to broken.

Are you suggesting to remove that callback parameter or to replace with a coro and run an `awaiting` of it ? I agree to respect the asyncio context, but in that case I don't understand a PITA with this callback. May I miss something ?

Moreother, `__aenter__` and `__aexit__`  methods  are missing in the Barrier class. To add with dedicated tests.

1st1 reviewed (changes requested):
I'm +1 for adding the primitive. But I think we should redesign the API. cc @ambv @asvetlov

YvesDup committed:
Update Doc/library/asyncio-sync.rst

Co-authored-by: Yury Selivanov <yury@edgedb.com>

YvesDup committed:
Update Doc/library/asyncio-sync.rst

Co-authored-by: Yury Selivanov <yury@edgedb.com>

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Add a complete documentation
Add a sample to show how use a Barrier Object

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Add a Barrier Class

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Add a BarrierTests class to run tests

YvesDup committed:
Merge branch 'python:main' into fix-issue-43352

YvesDup committed:
Merge branch 'python:main' into fix-issue-43352

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (pythonGH-24903)

New BarrierTests with 'IsolatedAsyncionTest`

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Comment corrections

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Update Barrier unit tests to use `unittest.IsolatedAsyncioTestCase`

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Update Barrier unit tests to use `unittest.IsolatedAsyncioTestCase`
--Missing newline at end of file--
--Erase bad comments--

YvesDup committed:
Merge branch 'python:main' into fix-issue-43352

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

remove `loop` parameter from `Barrier.__init__` method  (ref bpo-46796)

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Remove `test_barrier_doesnt_accept_loop_parameter` - Not necessary here, class Barrier is new
Rename 2 tests about 'abort' method

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

review tests
change unmeaning name of tests

YvesDup committed:
Merge branch 'python:main' into fix-issue-43352

asvetlov commented on Lib/asyncio/locks.py:
```diff
@@ -412,3 +412,215 @@ def release(self):
+    @property
+    def resetting(self):
+        """Return True if the barrier is resetting."""
+        return self._state == -1
```
Why all these attrs are exposed? Should user code use them?
For retrospection `__repr__` is enough.
In stdlib, much easier to add an attr on-demand than remove an existing one.

YvesDup replied:
1- When a waiting task is cancelled, there are two cases:
  * the task is the last waiting task: an `CancelledError` exception occurs, Barrier goes to broken state.
  * the task is waiting or blocked (draining state): this task quits the barrier which stays in the same state

Please see: `BarrierTests.test_filling_tasks_cancel_one` and `BarrierTests.test_draining_check_cancelled_action`

2- Ok, attrs are replaced with private methods, used in the `Barrier` class and tests.

asvetlov commented on Lib/asyncio/locks.py:
```diff
@@ -412,3 +412,215 @@ def release(self):
+            self._set_draining()
+            self._cond.notify_all()
+        except:
+            # an exception occurs during the _action coroutine.  
```
Wrap by `try/except` the only `await self._action()` call.
What is the expected behavior if `await action` raises CancelledError?

YvesDup replied:
Barrier goes to `broken`state and task goes to `cancelled`.  If an other error occurs in action, results are same
Please see `BarrierTests.test_draining_check_cancelled_action` and `Barrier.`test_draining_check_error_on_action`

asvetlov reviewed (commented):
What the behavior if a task waiting for the barrier is cancelled?

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Remove  broken, resetting, filling and draining attributes

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Add a new test `test_draining_check_cancelled_action`
Update `test_repr`

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Replace broken, resetting, filling and draining attributes with private methods
Update `__repr__`
Correction minor bugs on `_count_lock`attribut
Remove `_cout_wait` attribut
Refactoring code  on `_block()`method

asvetlov commented on Lib/asyncio/locks.py:
```diff
@@ -412,3 +412,218 @@ def release(self):
+            return self._count_block
+        return 0
+
+    def _set_broken(self):
```
There is a misunderstanding here.
I thought you keep `parties`, `n_waiting`, and `broken` attributes (they exist in `threading.Barrier` already).

All other attrs (including `n_waiting`) should not be exposed unless you can describe and protect an opposite opinion.

asvetlov replied:
Sorry if I did not express my intention clearly.

YvesDup replied:
> All other attrs (including n_waiting) should not be exposed unless you can describe and protect an opposite opinion.

Are you talking about `n_blocking`attr instead of `n_waiting` attr? I will remove it and restore `broken` attr.

> The cancellation behavior sounds reasonable.
> Please describe it in docs, the behavior is not very obvious.

Yes, I will do

> Sorry if I did not express my intention clearly.

It is ok for me, don't worry, your comments are very relevant

YvesDup replied:
Documentation is updated with cancellation task behavior.
`broken`attr is restored, `n_blocking` attr is removed.
Please @asvetlov could you approve running workflows. Thank

asvetlov reviewed (commented):
The cancellation behavior sounds reasonable.
Please describe it in docs, the behavior is not very obvious.

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Suppressed trailing white spaces

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Remove `n_blocking` attribut
Restore `broken` instead of `_is_broken()`
Rename all `_is_xxxx' methods to `_xxxx' where `xxxx' in draining, resetting, filling

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Use `_count_block` instead of removing `n_blocking` attr
Restore `broken` attr and use it instead of `_is_broken()` method
Rename all `_is_xxxx' methods to `_xxxx'.

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Description of the cancel behavior of tasks in the Barrier
Minor corrections

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Remove trailing spaces

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Suppress again trailing whitespaces :o

asvetlov reviewed (approved):
LGTM

1st1 commented on Doc/library/asyncio-sync.rst:
```diff
@@ -340,6 +341,153 @@ BoundedSemaphore
+   primitive for use by a fixed number of tasks that need to wait for each other.
+   Each of the tasks tries to pass the barrier by calling the :meth:`~Barrier.wait` method
+   and would block until all of the tasks have made their :meth:`~Barrier.wait` calls.
+   At this point, the tasks are released simultaneously.
```
```suggestion
   A barrier is a simple synchronization primitive that allows to block until a
   certain number of tasks are waiting on it.   
   Tasks can wait on the :meth:`~Barrier.wait` method and would be blocked until 
   the specified number of tasks end up waiting on :meth:`~Barrier.wait`.
   At that point all of the waiting tasks would unblock simultaneously.
```

1st1 commented on Doc/library/asyncio-sync.rst:
```diff
@@ -340,6 +341,153 @@ BoundedSemaphore
+   and would block until all of the tasks have made their :meth:`~Barrier.wait` calls.
+   At this point, the tasks are released simultaneously.
+
+   The barrier can be reused any number of times for the same number of tasks.
```
```suggestion
   The barrier can be reused any number of times.
```

1st1 commented on Doc/library/asyncio-sync.rst:
```diff
@@ -340,6 +341,153 @@ BoundedSemaphore
+          # Wait until folder+filenames are removed.
+          await asyncio.gather(*tasks)
+
+       asyncio.run(main("MyFolder", [f'file{i+1:02d}.txt' for i in range(NB)], NB))
```
This example is rather long and contrived. I'd drop it.

YvesDup replied:
Yes I agree. Do we need a simplier example ?

YvesDup replied:
I proposed an another example

1st1 commented on Doc/library/asyncio-sync.rst:
```diff
@@ -340,6 +341,153 @@ BoundedSemaphore
+   .. coroutinemethod:: wait()
+
+      Pass the barrier.  When all the tasks party to the barrier have called
+      this function, they are all released simultaneously.
```
```suggestion
      this function, they are all unblocked simultaneously.
```

1st1 commented on Doc/library/asyncio-sync.rst:
```diff
@@ -340,6 +341,153 @@ BoundedSemaphore
+
+      If an *action* was provided as a coroutine to the constructor, the last calling task of
+      :meth:`wait` method will have to *await this coroutine* prior to being released.
+      Should this call raise an error, including cancellation of task, the barrier is put into the broken state.
```
Long line, please trim to 79

YvesDup replied:
Done

1st1 commented on Lib/asyncio/locks.py:
```diff
@@ -412,3 +412,209 @@ def release(self):
+            # enter draining state
+            self._set_draining()
+            self._cond.notify_all()
+        except:
```
We shouldn't have any bare `except`. Replace with `except Exception`. Moreover, if this is to catch errors in `_action`, the `await self._action` should be surrounded with `try..except` and nothing else.

YvesDup replied:
done

1st1 commented on Lib/asyncio/locks.py:
```diff
@@ -412,3 +412,209 @@ def release(self):
+
+        # see if the barrier is in a broken state
+        if self.broken:
+            raise BrokenBarrierError
```
I think there should be a message explaining why specifically the barrier is broken now

YvesDup replied:
Done

1st1 commented on Lib/asyncio/locks.py:
```diff
@@ -412,3 +412,209 @@ def release(self):
+        await self._cond.wait_for(lambda: not self._filling())
+
+        if self.broken or self._resetting():
+            raise BrokenBarrierError
```
Add message

YvesDup replied:
Done

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Co-authored-by: Yury Selivanov <yury@edgedb.com>

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Co-authored-by: Yury Selivanov <yury@edgedb.com>

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Co-authored-by: Yury Selivanov <yury@edgedb.com>

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Proposition of a new example
Trailing whitespaces
Trim to 79 some lines

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Add messages in the BrokenBarrierError
Refactoring in the empty `except:` block in `of `_release()`
(Set list of exceptions, move except block)
Rewrite try/except block of `_block()`

1st1 commented on Doc/library/asyncio-sync.rst:
```diff
@@ -340,6 +341,132 @@ BoundedSemaphore
+      <asyncio.locks.Barrier object at 0x103e58bf0 [unlocked, state:0]>
+
+   As shown in this example, the :meth:`~Barrier.wait` method: can also be call with
+   :keyword:`async with` statement.
```
```suggestion
   The example also demonstrates using `async with` as 
   an alternative to awaiting on ``barrier.wait()``.
```

1st1 commented on Doc/library/asyncio-sync.rst:
```diff
@@ -340,6 +341,132 @@ BoundedSemaphore
+   As shown in this example, the :meth:`~Barrier.wait` method: can also be call with
+   :keyword:`async with` statement.
+
+   The barrier can be reused any number of times for the same number of tasks.
```
Drop this sentence, we already have it few lines earlier.

YvesDup replied:
Done

1st1 commented on Doc/library/asyncio-sync.rst:
```diff
@@ -340,6 +341,132 @@ BoundedSemaphore
+.. class:: Barrier(parties, action=None)
+
+   Create a barrier object for *parties* number of tasks.  A coroutine *action*,
+   when provided, is awaited once just before the release phasis (draining state).
```
```suggestion
   when provided, is awaited once just before the release occurs (draining state).
```

1st1 commented on Doc/library/asyncio-sync.rst:
```diff
@@ -340,6 +341,132 @@ BoundedSemaphore
+
+      When a task in the barrier is cancelled, two cases are possible:
+
+      * the task is the last waiting task, an :class:`BrokenBarrierError`
```
```suggestion
      * if the task is the last waiting task, a :class:`BrokenBarrierError`
```

1st1 commented on Doc/library/asyncio-sync.rst:
```diff
@@ -340,6 +341,132 @@ BoundedSemaphore
+      When a task in the barrier is cancelled, two cases are possible:
+
+      * the task is the last waiting task, an :class:`BrokenBarrierError`
+        exception occurs, barrier is put into a broken state;
```
```suggestion
        exception is raised. The barrier is put into a broken state;
```

1st1 commented on Doc/library/asyncio-sync.rst:
```diff
@@ -340,6 +341,132 @@ BoundedSemaphore
+      * the task is the last waiting task, an :class:`BrokenBarrierError`
+        exception occurs, barrier is put into a broken state;
+
+      * the task is waiting (filling state) or blocked (draining state),
```
```suggestion
      * if the task is waiting (filling state) or blocked (draining state),
```

1st1 commented on Doc/library/asyncio-sync.rst:
```diff
@@ -340,6 +341,132 @@ BoundedSemaphore
+
+      * the task is waiting (filling state) or blocked (draining state),
+        this task exits the barrier which stays in the same state.
+        If the state is filling, count of waiting task decreases by 1.
```
```suggestion
        If the state of the barrier is "filling", the number of waiting task decreases by 1.
```

1st1 commented on Doc/library/asyncio-sync.rst:
```diff
@@ -340,6 +341,132 @@ BoundedSemaphore
+        this task exits the barrier which stays in the same state.
+        If the state is filling, count of waiting task decreases by 1.
+
+      The return value is an integer in the range 0 to *parties-1*, different
```
```suggestion
      The return value is an integer in the range of 0 to ``parties-1``, different
```

1st1 commented on Lib/asyncio/locks.py:
```diff
@@ -412,3 +412,212 @@ def release(self):
+class Barrier(mixins._LoopBoundMixin):
+    """Asyncio equivalent to threading.Barrier
+
+    Implements a Barrier.
```
```suggestion
    Implements a Barrier primitive.
```

YvesDup replied:
done

1st1 commented on Lib/asyncio/locks.py:
```diff
@@ -412,3 +412,212 @@ def release(self):
+
+    def __init__(self, parties, action=None):
+        """Create a barrier, initialised to 'parties' tasks.
+        'action' is a coroutine which, when supplied, will be called by
```
```suggestion
        """Create a barrier, initialized to 'parties' tasks.
        
        'action' is an awaitable which, when supplied, will be called by
```

1st1 replied:
Actually, this is an interesting question -- a barrier can be reused multiple times, right?

If so, the `action` coroutine could be awaited multiple times too, right? If so, it should be `action_factory` and be a callable that returns an awaitable.

YvesDup replied:
Yes you are right. `action` must be a factory, a generator of coroutines. I wrote a test from the example of documentation. The `done()`coroutine can only be call once. :(
In the initial version of Barrier, I had used `action` as a function, not as a coroutine. 
So, this will be more complicated than expected !! 

But to be consistent with `threading.Barrier`, can we only use `action` as a callback function ? Or do we remove this parameter ? What do you think about now ? Thank

asvetlov replied:
I'm pretty sure that async code should use an async callback.
The `action` argument removal is an option. It could be added later on-demand, the addition is backward compatible.

YvesDup replied:
> I'm pretty sure that async code should use an async callback

Ok

I think there is a solution: `action` must be a `coroutinefunction`, even if it is included in a `functools.partial` call.  The type of `action` will be checked in the `__init__` method.
Now the `action` parameter is a callable, returning an awaitable. The call to `action` becomes `await self._action()`

What do you think of this ?

1st1 replied:
> I think there is a solution: action must be a coroutinefunction

It should be a function that returns an awaitable. An awaitable is a coroutine, a Task, or a Future object.

> Now the action parameter is a callable, returning an awaitable. The call to action becomes await self._action()

Yeah.

That said, maybe we should drop `action` for now. We can introduce it later if users want it. Implementing it in user code is almost trivial -- just do n+1 barrier with an extra task waiting on the barrier and running some code after the `with` block.

So I'm in favor of dropping this parameter for now.

asvetlov replied:
+1 for dropping.

Moreover, since the waiting order is deterministic in `asyncio.Condition`, the first awaited task wakes up first before all others.

YvesDup replied:
+1 for dropping.
I remove `action` parameter

YvesDup replied:
Should we move `BrokenBarrierError` exception from 'asyncio.locks.py'  to 'asyncio.exceptions.py' file ?

asvetlov replied:
> Should we move BrokenBarrierError exception...

I guess it would be nice but not strictly necessary.
`exceptions.py` reduces cross-imports. I expect that `BrokenBarrierError` is never used outside of the `Barrier` class itself.
On the other hand, putting all classes together could be good for source side readers.

YvesDup replied:
I 'd just seen you made modifications in python files. 
Could you give me time to read what you did please? Thank

asvetlov replied:
Yes, sure. 
No rush, please.

1st1 reviewed (commented):
Some nits and a question about `action`.

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)Update Doc/library/asyncio-sync.rst

Co-authored-by: Yury Selivanov <yury@edgedb.com>

YvesDup committed:
Update Doc/library/asybpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)ncio-sync.rst

Co-authored-by: Yury Selivanov <yury@edgedb.com>

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)Update Doc/library/asyncio-sync.rst

Co-authored-by: Yury Selivanov <yury@edgedb.com>

YvesDup committed:
Update Doc/library/asyncio-sync.rstbpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Co-authored-by: Yury Selivanov <yury@edgedb.com>

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Co-authored-by: Yury Selivanov <yury@edgedb.com>

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Co-authored-by: Yury Selivanov <yury@edgedb.com>

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Co-authored-by: Yury Selivanov <yury@edgedb.com>

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Suppress duplicated lines
Suppress trailing whitespaces

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Last corrections from @1st1
Modify type of action: coroutine to coroutinefunction

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Modifiy action type from coroutine to coroutinefunction
Add 2 dedicated tests

YvesDup committed:
po-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Fix example

YvesDup committed:
po-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Add a test to check multiple reuses of a barrier object.

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Remove `action` paramater

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Remove `import coroutines`
Suppress trailing whitespaces

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

minor correction

asvetlov committed:
Rewrite repr(), use enum for state, drop _count_blocked used by tests only

asvetlov committed:
Inline method

asvetlov committed:
Move BrokenBarrierError to exceptions.py

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Update sample with new `__repr__`

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Missing a newline

asvetlov committed:
Tune docs

YvesDup committed:
bpo-43352: Add a Barrier object to asyncio synchronized primitives (GH-24903)

Minor correction in example

asvetlov committed:
Bump
